### PR TITLE
Add support for fully qualified error codes in service client HTTP exceptions

### PR DESCRIPTION
**Pull request by timtay-microsoft**
Previously, the service client ignored the fully qualified status codes embedded in the error message for cases where the service threw a 404/500/etc.

Customer's need to have some granularity in treating a 404001 (deviceNotFound) from a 404010 (moduleNotFound) from a 404103 (deviceNotOnline) as these are all potential error cases when invoking a method on a module.

Now all of our service side exception types have dedicated fields for error code and error code description.

The error code will be the 6 digit code (404001, for instance) if the service gave the SDK that code, or it will be the 3 digit error code that we always get in the http response object (404, for instance)

### CODE REVIEW

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpSendHandler.java [hunk 1]
```diff
@@ -266,7 +266,15 @@ public void onDelivery(Event event)
...
-                log.debug("Closing amqp cloud to device message sender link since the message was delivered");
+                if (remoteState.getClass().equals(Accepted.class))
+                {
+                    log.debug("Closing AMQP cloud to device message sender link since the message was delivered");
+                }
+                else
+                {
+                    log.debug("Closing AMQP cloud to device message sender link since the message failed to be delivered");
+                }
+
```
Comment (timtay-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpSendHandler.java:277]: This change is completely orthogonal to the error code work that this PR is for, but it is a small change that is worth fixing despite that

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
+package com.microsoft.azure.sdk.iot.deps.serializer;
+
+/**
+ * See https://docs.microsoft.com/en-us/rest/api/iothub/common-error-codes for additional details
+ */
+public enum ErrorCodeDescription
```
Comment (abhipsaMisra) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:6]: This will be a good addition to the C# SDK as well

Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:6]: I was planning on doing the same over there, too, yeah

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            case (413):
+                return RequestEntityTooLarge;
+            case (429):
+                return ThrottlingException;
+            case (500):
+                return InternalServerError;
+            case (503):
+                return ServiceUnavailable;
+            default:
+                return UnclassifiedErrorCode;
```
Comment (prmathur-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:190]: Have these been validated by service ? i.e. do they support the above for what do they mean?

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothub/methods/DeviceMethodTests.java [hunk 4]
```diff
@@ -43,7 +44,7 @@ public void cleanToStart() throws Exception
     {
         super.cleanToStart();
     }
-
+/*
     @Test
```
Comment (prmathur-microsoft) [inline iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothub/methods/DeviceMethodTests.java:48]: Why is this commented?

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 3 above)
Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:190]: These are all taken directly from the C# sdk [here](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/iothub/service/src/Common/Exceptions/ErrorCode.cs)

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothub/methods/DeviceMethodTests.java (hunk 4 above)
Comment (timtay-microsoft) [reply iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothub/methods/DeviceMethodTests.java:48]: Whoops, my mistake. I can uncomment these now

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    /**
+     * Get the ErrorCodeDescription tied to the provided errorCode
+     * @param errorCode the service error code, such as 404, or 429001
+     * @return the corresponding ErrorCodeDescription
+     */
+    public static ErrorCodeDescription Parse(int errorCode)
+    {
+        switch (errorCode)
+        {
+            case(400001):
```
Comment (abhipsaMisra) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:85]: Add a link to the doc that has these mappings. Is it on some MSDN doc?

Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:85]: Unfortunately, I couldn't find a MSDN doc for this. These codes are taken directly from the C# SDK [here](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/iothub/service/src/Common/Exceptions/ErrorCode.cs)

Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:85]: In our work with them over the track 2 SDK, we can push for a document like that, though

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothub/methods/DeviceMethodTests.java [hunk 6]
```diff
@@ -43,7 +44,7 @@ public void cleanToStart() throws Exception
     {
         super.cleanToStart();
     }
-
+/*
```
Comment (abhipsaMisra) [inline iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothub/methods/DeviceMethodTests.java]: what is this for?

Comment (timtay-microsoft) [reply iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothub/methods/DeviceMethodTests.java]: I forgot to undo that, but now it is gone

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 3 above)
Comment (prmathur-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:190]: 👍

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    TooManyModulesOnDevice,
+    ThrottleBacklogLimitExceeded,
+    InvalidThrottleParameter,
+    ServerError,
+    JobCancelled,
+    ConnectionForcefullyClosedOnNewConnection,
+    DeviceNotOnline,
+    DeviceConnectionClosedRemotely,
+
+    // These descriptions belong to 3 digit http response codes such as 404.
```
Comment (abhipsaMisra) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:50]: nit: need consistency in using single-line vs block comments

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    Forbidden,
+    NotFound,
+    Conflict,
+
+    /**
+     * Represents status code for 429 and for status code 429001
+     */
+    PreconditionFailed,
+
+    RequestEntityTooLarge,
```
Comment (abhipsaMisra) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:62]: Does this comment mean PreconditionFailed maps to 429 and RequestEntityTooLarge to 429001?

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    /**
+     * Represents status code for 429 and 429001
+     */
+    ThrottlingException,
+    InternalServerError,
+
+    /**
+     * Defined for status code 503 and 503001
+     */
+    ServiceUnavailable;
```
Comment (abhipsaMisra) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:73]: Both of these codes map to ServiceUnavailable?

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 7 above)
Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:50]: The single line comments are for us developers to see. The block comments will show up when a user has intellisense

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java [hunk 10]
```diff
@@ -0,0 +1,193 @@
+package com.microsoft.azure.sdk.iot.deps.serializer;
+
+/**
+ * See https://docs.microsoft.com/en-us/rest/api/iothub/common-error-codes for additional details
+ */
+public enum ErrorCodeDescription
+{
+    // These descriptions belong to fully qualified status codes, such as 404001
+    UnclassifiedErrorCode,
```
Comment (abhipsaMisra) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:9]: Is there a better name for this enum than "Description"? This isn't really a description.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 8 above)
Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:62]: The javadoc is only visible when using intellisense for the PreconditionFailed enum value. It means that the service may return a 429, and it may return a 429001, but in both cases it maps to this enum value

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 10 above)
Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:9]: The enum is called "ErrorCodeDescription" because it describes what the error code (404001, etc) mean. I can't think of a better name.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 7 above)
Comment (abhipsaMisra) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:50]: really? I didn't know that!
So when I hover over `PreconditionFailed`, I should be seeing
```
/**
     * Represents status code for 429 and for status code 429001
     */
```

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 9 above)
Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:73]: Yep, weirdly

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 8 above)
Comment (abhipsaMisra) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:62]: Ah, the spacing and block comments confused me.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 7 above)
Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:50]: Without the slashes or stars, but yeah!

Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:50]: It's the java equivalent of how C# does 
```csharp
// Comment 
```
vs
```csharp
/// <summary>Comment</summary>
```

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 9 above)
Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:73]: Registry manager ops gets 503, whereas a device method invocation likely gets a 503001 based on my research so far. In both cases, though, the service is telling the SDK that the service is unavailable

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java (hunk 7 above)
Comment (timtay-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/ErrorCodeDescription.java:50]: Just to be more clear, though, if the block comments start with only one *, then they behave like the single line comments.

```java
/**
 * This is a javadoc header
 */

/*
 * This is just a comment. Users won't see it
 */
```
